[BUG] - 0.4 QHub render does not product .gitignore file to ignore terraform produced downloads

### OS system and architecture in which you are running QHub

Linux / AWS 

### Expected behavior

clean redeploy with github actions

### Actual behavior

remote: error: Trace: f7ee2b2e8ee572afea92b84d8b5108190bbd4a9a708bf89976c48eaa9ce39290        
remote: error: See http://git.io/iEPt8g for more information.        
remote: error: File stages/01-terraform-state/aws/.terraform/providers/registry.terraform.io/hashicorp/aws/3.73.0/linux_amd64/terraform-provider-aws_v3.73.0_x5 is 236.41 MB; this exceeds GitHub's file size limit of 100.00 MB        
remote: error: File stages/03-kubernetes-initialize/.terraform/providers/registry.terraform.io/hashicorp/aws/4.7.0/linux_amd64/terraform-provider-aws_v4.7.0_x5 is 238.45 MB; this exceeds GitHub's file size limit of 100.00 MB        
remote: error: GH001: Large files detected. You may want to try Git Large File Storage - https://git-lfs.github.com./        
To https://github.com/ESIPFed/qhub-maple
 ! [remote rejected] main -> main (pre-receive hook declined)
error: failed to push some refs to 'https://github.com/ESIPFed/qhub-maple'
Error: Process completed with exit code 1.

### How to Reproduce the problem?

not sure

### Command output

_No response_

### Versions and dependencies used.

_No response_

### Compute environment

_No response_

### Integrations

_No response_

### Anything else?

_No response_

--- a/.github/workflows/test.yaml
+++ b/.github/workflows/test.yaml
@@ -47,7 +47,7 @@ jobs:
         run: |
           black --version
           black qhub --diff --exclude "qhub/_version.py"
-          black --check qhub
+          black --check qhub --exclude "qhub/_version.py"
       - name: Flake8 Formatting
         run: |
           flake8 --version

--- a/.github/workflows/test.yaml
+++ b/.github/workflows/test.yaml
@@ -46,7 +46,7 @@ jobs:
       - name: Black Formatting
         run: |
           black --version
-          black qhub --diff --exclude qhub/_version.py
+          black qhub --diff --exclude "qhub/_version.py"
           black --check qhub
       - name: Flake8 Formatting
         run: |

--- a/.github/workflows/test.yaml
+++ b/.github/workflows/test.yaml
@@ -46,7 +46,7 @@ jobs:
       - name: Black Formatting
         run: |
           black --version
-          black qhub --diff
+          black qhub --diff --exclude qhub/_version.py
           black --check qhub
       - name: Flake8 Formatting
         run: |

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -7,7 +7,7 @@ repos:
     -   id: end-of-file-fixer
     -   id: trailing-whitespace
   - repo: https://github.com/psf/black
-    rev: 20.8b1
+    rev: 22.3.0
     hooks:
     -   id: black
         exclude: '/qhub/template/'

--- a/docs/ext/substitute.py
+++ b/docs/ext/substitute.py
@@ -3,14 +3,16 @@ This is a hard replace as soon as the source file is read, so no respect for any
 Simply forces replace of ||QHUB_VERSION|| with the qhub_version_string in conf.py
 """
 
+
 def dosubs(app, docname, source):
     """
     Replace QHUB_VERSION with the qhub version
     """
-    if app.config.qhub_version_string != '':
+    if app.config.qhub_version_string != "":
         src = source[0]
-        source[0] = src.replace('||QHUB_VERSION||', app.config.qhub_version_string)
+        source[0] = src.replace("||QHUB_VERSION||", app.config.qhub_version_string)
+
 
 def setup(app):
     app.connect("source-read", dosubs)
-    app.add_config_value('qhub_version_string', '', 'env')
+    app.add_config_value("qhub_version_string", "", "env")

--- a/environment-dev.yaml
+++ b/environment-dev.yaml
@@ -11,7 +11,7 @@ dependencies:
   - python-kubernetes
   # dev dependencies
   - flake8 ==3.8.4
-  - black ==20.8b1
+  - black ==22.3.0
   - twine
   - pytest
   - diagrams

--- a/qhub/render.py
+++ b/qhub/render.py
@@ -166,6 +166,21 @@ def render_contents(config: Dict):
     return contents
 
 
+def gen_gitignore(config):
+    toignore = """
+        # ignore terraform state
+        .terraform
+        terraform.tfstate
+        terraform.tfstate.backup
+        .terraform.tfstate.lock.info
+
+        # python
+        __pycache__
+    """
+
+    return {".gitignore": toignore}
+
+
 def gen_cicd(config):
     """
     Use cicd schema to generate workflow files based on the

--- a/qhub/render.py
+++ b/qhub/render.py
@@ -163,11 +163,20 @@ def render_contents(config: Dict):
                 }
             )
 
+    contents.update(gen_gitignore(config))
+
     return contents
 
 
 def gen_gitignore(config):
-    toignore = """
+    """
+    Generate `.gitignore` file.
+    Add files as needed.
+    """
+
+    from inspect import cleandoc
+
+    filestoignore = """
         # ignore terraform state
         .terraform
         terraform.tfstate
@@ -177,8 +186,7 @@ def gen_gitignore(config):
         # python
         __pycache__
     """
-
-    return {".gitignore": toignore}
+    return {".gitignore": cleandoc(filestoignore)}
 
 
 def gen_cicd(config):

--- a/qhub/stages/checks.py
+++ b/qhub/stages/checks.py
@@ -148,7 +148,7 @@ def check_ingress_dns(stage_outputs, config, disable_prompt):
 
     attempt = 0
     while not _attempt_dns_lookup(domain_name, ip):
-        sleeptime = 60 * (2 ** attempt)
+        sleeptime = 60 * (2**attempt)
         if not disable_prompt:
             input(
                 f"After attempting to poll the DNS, the record for domain={domain_name} appears not to exist, "

--- a/qhub/template/stages/07-kubernetes-services/modules/kubernetes/services/conda-store/config/conda_store_config.py
+++ b/qhub/template/stages/07-kubernetes-services/modules/kubernetes/services/conda-store/config/conda_store_config.py
@@ -13,9 +13,7 @@ c.CondaStore.database_url = "postgresql+psycopg2://${postgres-username}:${postgr
 c.CondaStore.default_uid = 1000
 c.CondaStore.default_gid = 100
 c.CondaStore.default_permissions = "775"
-c.CondaStore.conda_included_packages = [
-    "ipykernel"
-]
+c.CondaStore.conda_included_packages = ["ipykernel"]
 
 c.S3Storage.internal_endpoint = "${minio-service}:9000"
 c.S3Storage.internal_secure = False

--- a/qhub/template/stages/07-kubernetes-services/modules/kubernetes/services/jupyterhub/files/examples/dashboard_plotly.py
+++ b/qhub/template/stages/07-kubernetes-services/modules/kubernetes/services/jupyterhub/files/examples/dashboard_plotly.py
@@ -13,26 +13,27 @@ app = Dash(__name__)
 
 # assume you have a "long-form" data frame
 # see https://plotly.com/python/px-arguments/ for more options
-df = pd.DataFrame({
-    "Fruit": ["Apples", "Oranges", "Bananas", "Apples", "Oranges", "Bananas"],
-    "Amount": [4, 1, 2, 2, 4, 5],
-    "City": ["SF", "SF", "SF", "Montreal", "Montreal", "Montreal"]
-})
+df = pd.DataFrame(
+    {
+        "Fruit": ["Apples", "Oranges", "Bananas", "Apples", "Oranges", "Bananas"],
+        "Amount": [4, 1, 2, 2, 4, 5],
+        "City": ["SF", "SF", "SF", "Montreal", "Montreal", "Montreal"],
+    }
+)
 
 fig = px.bar(df, x="Fruit", y="Amount", color="City", barmode="group")
 
-app.layout = html.Div(children=[
-    html.H1(children='Hello Dash'),
-
-    html.Div(children='''
+app.layout = html.Div(
+    children=[
+        html.H1(children="Hello Dash"),
+        html.Div(
+            children="""
         Dash: A web application framework for your data.
-    '''),
+    """
+        ),
+        dcc.Graph(id="example-graph", figure=fig),
+    ]
+)
 
-    dcc.Graph(
-        id='example-graph',
-        figure=fig
-    )
-])
-
-if __name__ == '__main__':
+if __name__ == "__main__":
     app.run_server(debug=True)

--- a/qhub/template/stages/07-kubernetes-services/modules/kubernetes/services/jupyterhub/files/examples/dashboard_streamlit.py
+++ b/qhub/template/stages/07-kubernetes-services/modules/kubernetes/services/jupyterhub/files/examples/dashboard_streamlit.py
@@ -2,10 +2,9 @@ import streamlit as st
 import numpy as np
 import pandas as pd
 
-st.title('My first StreamLit QHub app')
+st.title("My first StreamLit QHub app")
 
 st.write("Here's our first attempt at using data to create a table:")
-st.write(pd.DataFrame({
-    'first column': [1, 2, 3, 4],
-    'second column': [10, 20, 30, 40]
-}))
+st.write(
+    pd.DataFrame({"first column": [1, 2, 3, 4], "second column": [10, 20, 30, 40]})
+)

--- a/qhub/template/stages/07-kubernetes-services/modules/kubernetes/services/jupyterhub/files/jupyterhub/02-spawner.py
+++ b/qhub/template/stages/07-kubernetes-services/modules/kubernetes/services/jupyterhub/files/jupyterhub/02-spawner.py
@@ -32,7 +32,7 @@ if cdsdashboards["enabled"]:
     c.CDSDashboardsConfig.spawn_default_options = False
 
     c.CDSDashboardsConfig.conda_envs = [
-        environment['name'] for _, environment in conda_store_environments.items()
+        environment["name"] for _, environment in conda_store_environments.items()
     ]
 else:
     c.JupyterHub.allow_named_servers = False

--- a/setup.cfg
+++ b/setup.cfg
@@ -40,7 +40,7 @@ install_requires =
 [options.extras_require]
 dev =
     flake8==3.8.4
-    black==20.8b1
+    black==22.3.0
     twine
     pytest
     pytest-timeout

--- a/tests/scripts/minikube-loadbalancer-ip.py
+++ b/tests/scripts/minikube-loadbalancer-ip.py
@@ -4,13 +4,13 @@ import os
 import sys
 import subprocess
 
-minikube_cmd = ['minikube', 'ssh', '--', 'ip', '-j', 'a']
+minikube_cmd = ["minikube", "ssh", "--", "ip", "-j", "a"]
 minikube_output = subprocess.check_output(minikube_cmd, encoding="utf-8")[:-1]
 
 address = None
 for interface in json.loads(minikube_output):
-    if interface['ifname'] == 'eth0':
-        address = interface['addr_info'][0]['local'].split('.')
+    if interface["ifname"] == "eth0":
+        address = interface["addr_info"][0]["local"].split(".")
         break
 else:
     print("minikube interface eth0 not found")

--- a/tests/test_render.py
+++ b/tests/test_render.py
@@ -1,9 +1,11 @@
+import os
 import pytest
+from pathlib import Path
 
 from ruamel.yaml import YAML
 
 from qhub.render import render_template, set_env_vars_in_config
-from .conftest import render_config_partial
+from .conftest import render_config_partial, PRESERVED_DIR
 
 
 @pytest.fixture
@@ -87,3 +89,31 @@ def test_render_template(write_qhub_config_to_file):
     assert qhub_config_json["namespace"] == namespace
     assert qhub_config_json["domain"] == domain
     assert qhub_config_json["provider"] == cloud_provider
+
+
+def test_exists_after_render(write_qhub_config_to_file):
+    items_to_check = [
+        ".gitignore",
+        "image",
+        "stages",
+        "qhub-config.yaml",
+        PRESERVED_DIR,
+    ]
+
+    qhub_config_loc, _ = write_qhub_config_to_file
+
+    yaml = YAML()
+    qhub_config_json = yaml.load(qhub_config_loc.read_text())
+
+    # list of files/dirs available after `qhub render` command
+    ls = os.listdir(Path(qhub_config_loc).parent.resolve())
+
+    cicd = qhub_config_json.get("ci_cd", {}).get("type", None)
+
+    if cicd == "github-actions":
+        items_to_check.append(".github")
+    elif cicd == "gitlab-ci":
+        items_to_check.append(".gitlab-ci.yml")
+
+    for i in items_to_check:
+        assert i in ls

--- a/tests_deployment/test_jupyterhub_ssh.py
+++ b/tests_deployment/test_jupyterhub_ssh.py
@@ -1,4 +1,3 @@
-
 import re
 import uuid
 

